src/libultra/io/pfsreadwritefile.c

```diff
@@ -1,5 +1,124 @@
+#include "ultra64.h"
 #include "global.h"
 
-#pragma GLOBAL_ASM("asm/non_matchings/boot/pfsreadwritefile/__osPfsGetNextPage.s")
+#define CHECK_IPAGE(p, pfs)                                                                                   \
+    (((p).ipage >= (pfs).inodeStartPage) && ((p).inode_t.bank < (pfs).banks) && ((p).inode_t.page >= 0x01) && \
+     ((p).inode_t.page < 0x80))
 
-#pragma GLOBAL_ASM("asm/non_matchings/boot/pfsreadwritefile/osPfsReadWriteFile.s")
+s32 __osPfsGetNextPage(OSPfs* pfs, u8* bank, __OSInode* inode, __OSInodeUnit* page) {
+    s32 ret;
+
+    if (page->inode_t.bank != *bank) {
+        *bank = page->inode_t.bank;
+        if ((ret = __osPfsRWInode(pfs, inode, PFS_READ, *bank)) != 0) {
+            return ret;
+        }
+    }
+    *page = inode->inodePage[page->inode_t.page];
+
+    if (!CHECK_IPAGE(*page, *pfs)) {
+        if (page->ipage == PFS_EOF) {
+            return PFS_ERR_INVALID;
+        }
+        return PFS_ERR_INCONSISTENT;
+    }
+    return 0;
+}
+
+s32 osPfsReadWriteFile(OSPfs* pfs, s32 fileNo, u8 flag, s32 offset, s32 size, u8* data) {
```

Comment:
```suggestion
s32 osPfsReadWriteFile(OSPfs* pfs, s32 fileNo, u8 flag, s32 offset, size_t size, u8* data) {
```

Reply:
If I make this change, several places don't match anymore. In this function it specifically checks that size is greater than 0, so maybe they didn't actually use u32/size_t?